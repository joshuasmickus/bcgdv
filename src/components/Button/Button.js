import React from 'react';
import { func, node, oneOfType, string } from 'prop-types';


export const Button = ({ className, onClick, children }) => (
  <button className={`button ${className}`} onClick={onClick}>{children}</button>
);


Button.propTypes = {
  className: string,
  onClick: func.isRequired,
  children: oneOfType([
    string,
    node
  ]).isRequired
};

Button.defaultProps = {
  className: ''
}
