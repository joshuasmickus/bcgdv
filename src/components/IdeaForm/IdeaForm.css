.idea-form {
  border: 1px solid black;
  -moz-box-sizing:border-box;
  -webkit-box-sizing:border-box;
  box-sizing: border-box;
  display: inline-block;
  font-size: 1rem;
  height: 150px;
  margin: 10px;
  width: 150px;
  list-style: none;
  padding: 8px;
}

.idea-form label {
  display: block;
  padding: 8px 0;
}

.idea-form input {
  border: 0;
  width: 120px;
}

.idea-form input:focus {
 border: 1px solid lightgray;
 outline: 0;
}
