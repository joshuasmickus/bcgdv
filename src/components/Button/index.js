import './Button.css';

export { Button as default } from './Button';
