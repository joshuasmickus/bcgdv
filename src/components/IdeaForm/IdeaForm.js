import React, { Component } from 'react';
import { func, shape } from 'prop-types';

import { ideaShape } from '../../Constants/PropTypes';
import { finishedEditing, updateIdea } from '../../redux/ideas';

import Button from '../Button';

export class IdeaForm extends Component {
  state = {
    title: this.props.idea.title || '',
    body: this.props.idea.body || ''
  }

  handleChange = ({ target: { name, value } }) => {
    this.setState({ [name]: value});
  }

  handleBlur = (event) => {
    event.preventDefault();

    const { createdDate, id } = this.props.idea;
    const { body, title } = this.state;

    this.props.actions.updateIdea({ body, createdDate, id, title });
  }

  handleSubmit = (event) => {
    event.preventDefault();

    this.props.actions.finishedEditing();
  }

  render() {
    const { body, title } = this.state;

    return (
      <form className="idea-form">
        <label>Title: </label>
        <input
          autoFocus
          type="text"
          name="title"
          value={title}
          onBlur={this.handleBlur}
          onChange={this.handleChange}
        />
        <label>Body: </label>
        <input
          type="text"
          name="body"
          value={body}
          onBlur={this.handleBlur}
          onChange={this.handleChange}
          maxLength="140"
        />
        <Button onClick={this.handleSubmit}>Done</Button>
      </form>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  actions: {
    finishedEditing: (idea) => dispatch(finishedEditing.request()),
    updateIdea: (idea) => dispatch(updateIdea.request(idea))
  }
})

IdeaForm.propTypes = {
  actions: shape({
    finishedEditing: func.isRequired,
    updateIdea: func.isRequired
  }).isRequired,
  idea: shape(ideaShape)
};
