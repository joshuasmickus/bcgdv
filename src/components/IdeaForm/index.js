import { connect } from 'react-redux';

import { IdeaForm, mapDispatchToProps } from './IdeaForm';

import './IdeaForm.css';

export default connect(null, mapDispatchToProps)(IdeaForm);
