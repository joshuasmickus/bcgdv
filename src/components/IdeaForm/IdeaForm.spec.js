import React from 'react';
import { shallow } from 'enzyme';

import Button from '../Button';

import { IdeaForm } from './IdeaForm';

const finishedEditingMock = jest.fn();
const updateIdeaMock = jest.fn();
const actions = {
  finishedEditing: finishedEditingMock,
  updateIdea: updateIdeaMock
};
const idea = {
  id: 'something',
  createdDate: 'today',
  title: 'my idea',
  body: 'something under 140 chars'
};

describe('Given an IdeaForm component', () => {
  describe('when it is rendered', () => {
    beforeEach(() => {
      finishedEditingMock.mockReset();
      updateIdeaMock.mockReset();
    });

    it('should match the snapshot', () => {
      const wrapper = shallow(<IdeaForm actions={actions} idea={idea} />);

      expect(wrapper.getElement()).toMatchSnapshot();
    });

    describe('and the done button is clicked', () => {
      it('should trigger finishedEditing action', () => {
        const preventDefaultMock = jest.fn();
        const wrapper = shallow(<IdeaForm actions={actions} idea={idea} />);

        wrapper.find(Button).simulate('click', { preventDefault: preventDefaultMock });

        expect(finishedEditingMock).toHaveBeenCalledWith();
        expect(preventDefaultMock).toHaveBeenCalledWith();
      });
    });
  });
});
