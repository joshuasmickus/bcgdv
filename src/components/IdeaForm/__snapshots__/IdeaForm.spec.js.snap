// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given an IdeaForm component when it is rendered should match the snapshot 1`] = `
<form
  className="idea-form"
>
  <label>
    Title: 
  </label>
  <input
    autoFocus={true}
    name="title"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
    value="my idea"
  />
  <label>
    Body: 
  </label>
  <input
    maxLength="140"
    name="body"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
    value="something under 140 chars"
  />
  <Button
    className=""
    onClick={[Function]}
  >
    Done
  </Button>
</form>
`;
